module/os/linux/zfs/zfs_racct.c

```diff
@@ -25,14 +25,35 @@
 
 #include <sys/zfs_racct.h>
 
+#ifdef _KERNEL
+#include <linux/task_io_accounting_ops.h>
+
+void
+zfs_racct_read(spa_t *spa, uint64_t size, uint64_t iops, uint32_t flags)
+{
+	task_io_account_read(size);
+	spa_iostats_read_add(spa, size, iops, flags);
+}
+
 void
-zfs_racct_read(uint64_t size, uint64_t iops)
+zfs_racct_write(spa_t *spa, uint64_t size, uint64_t iops, uint32_t flags)
 {
-	(void) size, (void) iops;
+	task_io_account_write(size);
+	spa_iostats_write_add(spa, size, iops, flags);
 }
 
+#else
+
 void
-zfs_racct_write(uint64_t size, uint64_t iops)
+zfs_racct_read(spa_t *spa, uint64_t size, uint64_t iops, uint32_t flags)
 {
-	(void) size, (void) iops;
+	(void) spa, (void) size, (void) iops, (void) flags;
 }
+
+void
+zfs_racct_write(spa_t *spa, uint64_t size, uint64_t iops, uint32_t flags)
+{
+	(void) spa, (void) size, (void) iops, (void) flags;
```

Comment:
So O_DIRECT only happen in kernel space. However, if you think it would be worth user space accounting, we could look into it. Just not sure it is worth it if I am understanding the suggestion correct.

Reply:
Never mind. I've forgot it is not used in user-space.  Though I wonder if user-space could simulate it for testing purposes.

Reply:
So at one point I did hook up user space stuff for the ABD alloc from pages stuff. However, that was all ripped out as it was overkill in our minds back in the day. I don't think this would be necessary for user space if I am being honest. This is a strictly a kernel thing. However, if you have thoughts of why this should be somehow tested in user space, let me know.